layer2-state-provider/history_commitment_provider.go

```diff
@@ -251,11 +253,9 @@ func (p *HistoryCommitmentProvider) historyCommitmentImpl(
 				rawStepHeights += ","
 			}
 		}
-		panic("TODO: JsonCollectMachineHashes FromState")
-		var zero uint64
 		collectMachineHashes := api.JsonCollectMachineHashes{
 			WasmModuleRoot:       cfg.WasmModuleRoot,
-			FromBatch:            1 / zero,
+			FromBatch:            cfg.FromState.Batch,
```

Comment:
Don't we also need to add FromPosInBatch here? I don't really know where JsonCollectMachineHashes is used

Reply:
Yeah. These API's are confusing. I *think* the `api.JsonCollectMachineHashes` is essentially only written and read from a database to track the progress of a long-running RPC call to collect the machine hashes. In the `bold` repository, it is read when trying to serve a response to a request to CollectMachineHashes. Since it has a pointer to finish time. I suppose someone could use the responses to know whether or not a particular set of hashes has been completely collected yet or not.

I *thought* it is okay not to have `PosInBatch` because the `BlockChallengeHeight` has the same information. But, now I'm not so sure about that. I see, in your changes to the implementation of `CollectMachineHashes` in the `nitro` repo, you are adding both values to the `prevBatchMessageCount` variable. So, maybe I don't understand what the variables are supposed to mean.

I think there are several sources of complexity/confusion we need to iron out and simplify:
1. At each challenge phase (i.e. Block, BigStep, SmallStep, OneStep) what data does BoLD need to pass to the nitro node to get the correct hashes for building the history commitment. Alternatively, how are we addressing the machine states from which we want to have the nitro node validate the states? (For example, batch-relative vs. absolute block heights.)
2. Do both the BoLD and nitro components need to read from the sequencer inbox?
3. How does the protocol change when dealing with assertions which overflow the max number of blocks, and how does it change when dealing with the assertions which follow those assertions (to ensure the right starting and ending addresses?)

In particular, I'm increasingly curious as to why the BoLD components even need to communicate about inbox batches. Isn't the range of L2Block numbers knowable from the assertion history and sufficient to communicate to the nitro node? (e.g. "Calculate a history commitment from block 7568 - block 9645" it doesn't matter which batches the corresponding messages are in.) Or, as another example, "Calculate the 2^23 hashes from block 8000, opcode index (8*(2^23)+1)".

Maybe I need to study the fields on the representation of an Assertion in the contracts more closely.

Reply:
Yeah. Okay. So, after looking at: https://github.com/OffchainLabs/nitro-contracts/blob/109a8a36cd4c6a2a0d2b5003b01adee60d83e2a1/src/state/GlobalState.sol#L53

I can see that the AssertionInfo has a BeforeState and AfterState which are both of type GlobalState and that the GlobalState always addresses the corresponding L2Block as the "InboxPosition" and "PositionInMessage". And now, we're passing an `AssertionMetadata.FromState` which is also a `GlobalState` and the nitro node has to translate that into an actual L2Block number by looking up the previous batch's highest message index, and adding the offset into that batch to get the actual block number.

Okay. I think this answers the question of "Do we want the BoLD side to need to read the inbox?" with "No."

Reply:
Yeah. I think that `BlockChallengeHeight` has the same infomration as `FromState.PosInBatch`.
So, maybe the thing to do is to take the `BlockChallengeHeight` off of the `HashCollectorConfig` and just use `FromState` to convey both pieces of data.
I can go ahead and try that. For now, I'll leave the name of the Json field `BlockChallengeHeight`, but we may decide that we want to serialize the `FromState` directly into the Json representation to get rid of the confusion.

Reply:
One thing that adds to the confusion here is that for a `HistoryCommitmentRequest`, if there `UpperChallengeOriginHeights` has a `len >= 1` then, I *think* that `UpperChallengeOriginHeights[0]` actually has the same value as `AssertionMetadata.FromState.PosInBatch` (even though that isn't explicitly checked anywhere. I believe this invariant is true because of two things:
1. All of the tests pass the same value when there is a value for `UpperChallengeOriginHeights[0]`
2. I just think that they capture the same data. How far into the specified `FromBatch` are we indexing to get to the message that corresponds to the L2 block for which we want to calculate a history commitment.

Reply:
Okay, the JsonCollectMachineHashes struct now has the PostionInBatch field as well.